utils/src/test/java/software/amazon/awssdk/utils/cache/lru/LruCacheTest.java

```diff
@@ -0,0 +1,372 @@
+        allExecutions = new ArrayList<>();
+    }
+
+
+    @AfterEach
+    public void shutdown() {
+        executorService.shutdown();
+    }
+
+    @Test
+    void when_cacheHasMiss_ValueIsCalculatedAndCached() {
+        LruCache<Key, Value> cache = simpleCache.get();
+        populateAndVerify(cache, 1);
+    }
+
+    @Test
+    void when_cacheHasHit_ValueIsRetrievedFromCache() {
+        LruCache<Key, Value> cache = simpleCache.get();
+
+        populateAndVerify(cache, MAX_SIMPLE_CACHE_SIZE);
+        Value result = cache.get(simpleTestKeys.get(2));
+        assertThat(cache.size()).isEqualTo(MAX_SIMPLE_CACHE_SIZE);
+        assertThat(result).isNotNull();
+        assertThat(result).isEqualTo(simpleTestValues.get(2));
+
+        cache.get(simpleTestKeys.get(2));
+
+        int onlyOneInvocation = 1;
+        verify(simpleValueSupplier, times(onlyOneInvocation)).apply(simpleTestKeys.get(2));
+    }
+
+    @Test
+    void when_cacheFillsUp_ValuesAreEvictedFromCache() {
+        LruCache<Key, Value> cache = simpleCache.get();
+
+        populateAndVerify(cache, MAX_SIMPLE_CACHE_SIZE);
+
+        //evict 1
+        Value result = cache.get(simpleTestKeys.get(4));
+        assertCacheState(cache, result, MAX_SIMPLE_CACHE_SIZE, 4);
+
+        //move 3 up
+        result = cache.get(simpleTestKeys.get(3));
+        assertCacheState(cache, result, MAX_SIMPLE_CACHE_SIZE, 3);
+
+        //evict 2
+        result = cache.get(simpleTestKeys.get(5));
+        assertCacheState(cache, result, MAX_SIMPLE_CACHE_SIZE, 5);
+
+        //move 4 up
+        result = cache.get(simpleTestKeys.get(4));
+        assertCacheState(cache, result, MAX_SIMPLE_CACHE_SIZE, 4);
+
+        //evict 3
+        result = cache.get(simpleTestKeys.get(2));
+        assertCacheState(cache, result, MAX_SIMPLE_CACHE_SIZE, 2);
+
+        verify(simpleValueSupplier, times(1)).apply(simpleTestKeys.get(1));
+        verify(simpleValueSupplier, times(2)).apply(simpleTestKeys.get(2));
+        verify(simpleValueSupplier, times(1)).apply(simpleTestKeys.get(3));
+        verify(simpleValueSupplier, times(1)).apply(simpleTestKeys.get(4));
+        verify(simpleValueSupplier, times(1)).apply(simpleTestKeys.get(5));
+    }
+
+    @Test
+    void when_multipleCallers_oneCallerBlocks_simple() throws InterruptedException {
+        ExecutorService executor = Executors.newCachedThreadPool();
+        try  {
+            LruCache<Key, Value> cache = simpleCache.get();
+
+            for (int i = 0; i < 1000; i++) {
+                int numEntry = ThreadLocalRandom.current().nextInt(MAX_SIMPLE_TEST_ENTRIES - 1);
+                executor.submit(() -> cache.get(simpleTestKeys.get(numEntry)));
+            }
+
+            executor.shutdown();
+            assertThat(executor.awaitTermination(10, TimeUnit.SECONDS)).isTrue();
+        } finally {
+            executor.shutdownNow();
+        }
+    }
+
+    @Test
+    void when_multipleCallers_oneCallerBlock_doesnt_Work() {
```

Comment:
I don't understand the test name (can't tell what it's supposed to be testing).

Reply:
It didn't work, and has been removed.